xformers/components/attention/csrc/cuda/attention.cu

```diff
@@ -1141,14 +1266,184 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> attention_backward(
         key,
         value,
         logsumexp,
-        tmp_sum_i);
+        output,
+        attn_bias,
+        p,
+        rng_engine_inputs);
   }
 
   AT_CUDA_CHECK(cudaGetLastError());
 
   return std::make_tuple(grad_q, grad_k, grad_v);
 }
 
+// the functions below are only used for testing
+// there is a lot of repetition compared to
+// the forward code, so this could be refactored
+// in the future
+
+template <
+    bool first,
+    typename scalar_t,
+    typename vec_t,
+    int kBlockSizeK,
+    int kBlockSizeQ,
+    int WARP_SIZE>
+struct UnrollLoopForMask {
+  static __device__ __forceinline__ void eval(
+      scalar_t* output[kBlockSizeQ],
+      int64_t N,
+      int64_t M,
+      at::PhiloxCudaState philox_args,
+      int64_t global_offset,
+      scalar_t p) {
+    constexpr int64_t step = kBlockSizeK * WARP_SIZE;
+    int64_t l;
+    if (first) {
+      l = threadIdx.x * kBlockSizeK;
+    } else {
+      l = N - (N & (2 * step - 1)) + threadIdx.x * kBlockSizeK;
+    }
+    // this is equivalent to N - N % step, but faster
+    // guaranteed to be the same as step is a power of 2
+    int64_t end_iter = kBlockSizeK == 1 ? N : N - (N & (step - 1));
+    scalar_t s_delta[kBlockSizeQ][kBlockSizeK];
+    int64_t query_idx =
+        blockIdx.x * (blockDim.y * kBlockSizeQ) + threadIdx.y * kBlockSizeQ;
+    // if (l < end_iter) {
+    {
+      for (; l < end_iter; l += step) {
+        for (int jj = 0; jj < kBlockSizeQ; jj++) {
```

Comment:
I think that I wrote the same in the previous PR related to that @fmassa, but it feels crazy that this is the best cuda has to offer to init s_delta.. I remember you wrote that was the gist of it, not really a question here, just thinking out loud